src/protocols/bdx/BDXMessageUtils.cpp

```diff
@@ -0,0 +1,408 @@
+
+using namespace chip;
+using namespace chip::BDX;
+using namespace chip::Encoding::LittleEndian;
+
+CHIP_ERROR TransferInit::Pack(System::PacketBuffer & aBuffer)
+{
+    CHIP_ERROR err   = CHIP_NO_ERROR;
+    uint8_t rangeCtl = 0;
+    uint8_t ptcByte  = 0;
+
+    BufBound bbuf(aBuffer.Start(), aBuffer.AvailableDataLength());
+
+    ptcByte |= mSupportedVersions & VERSION_MASK;
+    if (mSupportsSenderDrive)
+        ptcByte |= SENDER_DRIVE_MASK;
+    if (mSupportsReceiverDrive)
+        ptcByte |= RECEIVER_DRIVE_MASK;
+    if (mSupportsAsync)
+        ptcByte |= ASYNC_MASK;
+
+    if (mDefLen)
+        rangeCtl |= DEFLEN_MASK;
+    if (mStartOffset)
+        rangeCtl |= START_OFFSET_MASK;
+    if (mWideRange)
+        rangeCtl |= WIDERANGE_MASK;
+
+    bbuf.Put(ptcByte);
+    bbuf.Put(rangeCtl);
+    bbuf.PutLE16(mMaxBlockSize);
+
+    if (mStartOffset > 0)
+    {
+        if (mWideRange)
+        {
+            bbuf.PutLE64(mStartOffset);
+        }
+        else
+        {
+            bbuf.PutLE32(static_cast<uint32_t>(mStartOffset));
+        }
+    }
+
+    if (mDefLen)
+    {
+        if (mWideRange)
+        {
+            bbuf.PutLE64(mMaxLength);
+        }
+        else
+        {
+            bbuf.PutLE32(static_cast<uint32_t>(mMaxLength));
+        }
+    }
+
+    VerifyOrExit(mFileDesignator && mFileDesLength > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+    bbuf.PutLE16(mFileDesLength);
+    bbuf.Put(mFileDesignator, static_cast<size_t>(mFileDesLength));
+
+    // Metadata is optional
+    if (mMetadata && mMetadataLength > 0)
+    {
+        bbuf.Put(mMetadata, static_cast<size_t>(mMetadataLength));
+    }
+
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+    aBuffer.SetDataLength(static_cast<uint16_t>(bbuf.Written()));
+
+exit:
+    return err;
+}
+
+CHIP_ERROR TransferInit::Parse(System::PacketBuffer & aBuffer, TransferInit & aParsedMessage)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    uint8_t ptcByte;
+    uint8_t rangeCtl;
+    bool hasStartOffset     = false;
+    uint32_t tmpUint32Value = 0;
+    uint8_t * bufStart      = aBuffer.Start();
+    Reader bufReader(bufStart, aBuffer.DataLength());
+
+    SuccessOrExit(bufReader.Read8(&ptcByte).Read8(&rangeCtl).Read16(&aParsedMessage.mMaxBlockSize).StatusCode());
+
+    aParsedMessage.mSupportedVersions     = ptcByte & VERSION_MASK;
+    aParsedMessage.mSupportsSenderDrive   = ((ptcByte & SENDER_DRIVE_MASK) != 0);
+    aParsedMessage.mSupportsReceiverDrive = ((ptcByte & RECEIVER_DRIVE_MASK) != 0);
+    aParsedMessage.mSupportsAsync         = ((ptcByte & ASYNC_MASK) != 0);
+
+    aParsedMessage.mDefLen    = (rangeCtl & DEFLEN_MASK) != 0;
+    hasStartOffset            = (rangeCtl & START_OFFSET_MASK) != 0;
+    aParsedMessage.mWideRange = (rangeCtl & WIDERANGE_MASK) != 0;
+
+    if (hasStartOffset)
+    {
+        if (aParsedMessage.mWideRange)
+        {
+            SuccessOrExit(bufReader.Read64(&aParsedMessage.mStartOffset).StatusCode());
```

Comment:
you can do statuscode at the end and optimistically read everything. Code will look more linear at the cost of more processing in the error case (which should be rare).

Reply:
Good point. I think this would help readability, keeping the lines with Read() shorter

Reply:
and I think regardless, if there's an error in reading the message, the caller probably doesn't care exactly what error that was. It will just report that there was some error reading the message

Reply:
Ok so I tried this and then realized that Read() is declared with `CHECK_RETURN_VALUE` so I have to keep the checks

Reply:
I think that is an oversight since StatusCode() already exists. We can drop the check instead.

Reply:
It's not an oversight; it's purposeful.  If we don't do that, it's too easy for people to forget to do the `StatusCode()` call later.

One thing we _could_ consider doing is adding a `WillCheckStatusLater()` (with a better name?) that just returns `void` and is not `CHECK_RETURN_VALUE` (obviously).  Then you can write:
```
  bufReader.Read64(&aParsedMessage.mStartOffset).WillCheckStatusLater();
```
And as long as nothing between here and the status check uses the value of `aParsedMessage.mStartOffset` (which will then have to be enforced via code audit, I guess) it will be ok...